qiskit/synthesis/multi_controlled/mcx_synthesis.py

```diff
@@ -230,3 +233,116 @@ def synth_mcx_1_clean_b95(num_ctrl_qubits: int):
     )
 
     return qc
+
+
+def synth_mcx_gray_code(num_ctrl_qubits: int) -> QuantumCircuit:
+    r"""
+    Synthesize a multi-controlled X gate with :math:`k` controls using the Gray code.
+
+    Produces a quantum circuit with :math:`k + 1` qubits. This method
+    produces exponentially many CX gates and should be used only for small
+    values of :math:`k`.
+
+    Args:
+        num_ctrl_qubits: The number of control qubits.
+
+    Returns:
+        The synthesized quantum circuit.
+    """
+    num_qubits = num_ctrl_qubits + 1
+    q = QuantumRegister(num_qubits, name="q")
+    qc = QuantumCircuit(q, name="mcx_gray")
+    qc._append(HGate(), [q[-1]], [])
+    qc._append(MCU1Gate(np.pi, num_ctrl_qubits=num_ctrl_qubits), q[:], [])
```

Comment:
It would be more straightforward to just call `_gray_code_chain` with `gate=CXGate()` from `../library/u3.py` (and less confusing IMO since we have very different algorithms for `MCPhase` and `MCU1` even they should be literally the same) -- should we just update this while we're at it?

Reply:
Right now we have just moved the code without doing any real changes, for simplicity let's keep it this way for this PR. But I completely agree that we need to rethink the hierarchy of all the different multi-controlled synthesis methods available.